### ISSUE TITLE
Help wanted: Using L2CAP connection to read from and write settings to AirPods

### ISSUE THREAD

**Issue by d4rken** [d4rken-org/capod#215]
Our sister project for Windows, [MagicPods](https://github.com/steam3d/MagicPodsCore), has reverse engineered reading and writing data to AirPods.

* The good news is that this actually works and I could reproduce it on a desktop computer running linux. So support for changing settings (ANC mode, microphone etc.) and getting more reliable information would be possible

* The bad news is that I'm unable to get the connection established on an Android device.

**If anyone can figure out how to establish the connection, then I could start on adding support for this.**

---

AirPods require a Bluetooth connection that uses the `L2CAP` protocol. I was unable to establish this connection on a Pixel 6 running Android 15 and Pixel 8 running Android 14.

This is what I tried so far

Use the public `createInsecureL2capChannel` method

```
06:50:53.800 bt_l2cap                             E  packages/modules/Bluetooth/system/stack/l2cap/l2c_api.cc:417 - L2CA_RegisterLECoc: Invalid BLE PSM value, PSM: 0x1001
06:50:53.800 bt_stack                             E  [ERROR:gap_conn.cc(255)] GAP_ConnOpen: Failure registering PSM 0x1001
```
but that seems to have limits on the `PSM` value that you can provide

https://cs.android.com/android/platform/superproject/main/+/main:packages/modules/Bluetooth/system/stack/l2cap/l2c_api.cc;l=416?q=%22Invalid%20BLE%20PSM%20value%22 

```
#define L2C_IS_VALID_LE_PSM(psm) (((psm) > 0x0000) && ((psm) < 0x0100))
```

Then I tried the hidden method `createInsecureL2capSocket`

https://cs.android.com/android/platform/superproject/main/+/main:packages/modules/Bluetooth/framework/java/android/bluetooth/BluetoothDevice.java;l=2924?q=createInsecureL2capSocket&sq=&ss=android%2Fplatform%2Fsuperproject%2Fmain

> Accessing hidden method Landroid/bluetooth/BluetoothDevice;->createInsecureL2capSocket(I)Landroid/bluetooth/BluetoothSocket; (max-target-o, reflection, denied)

Which needed some convincing to execute (https://github.com/LSPosed/AndroidHiddenApiBypass) but

```
07:44:04.151 bluetooth                            W  packages/modules/Bluetooth/system/stack/l2cap/l2c_fcr.cc:1603 - l2c_fcr_chk_chan_modes: L2CAP - Peer does not support our desired channel types
07:44:04.151 bt_btif_sock                         I  packages/modules/Bluetooth/system/btif/src/btif_sock.cc:164 - btif_sock_connection_logger: address=xx:xx:xx:xx:e4:0a, state=4, role=2, server_name=, channel=4097
07:44:04.151 bluetooth                            I  packages/modules/Bluetooth/system/btif/src/btif_sock_l2cap.cc:363 - send_app_err_code: Sending l2cap failure reason socket_id:34 reason code:1
07:44:04.151 bt_btif_sock                         I  packages/modules/Bluetooth/system/btif/src/btif_sock.cc:164 - btif_sock_connection_logger: address=xx:xx:xx:xx:e4:0a, state=5, role=2, server_name=, channel=4097
07:44:04.151 bluetooth                            I  packages/modules/Bluetooth/system/btif/src/btif_sock_l2cap.cc:234 - btsock_l2cap_free_l: Application has already closed l2cap socket socket_id:34

android.bluetooth.BluetoothSocketException: Connection failed for unknown reason 
```

still no dice.

A friend has tried simulating AirPods on a Desktop computer and was able to connect to that mock via `createInsecureL2capSocket` from an Android phone, but wasn't able to do this with actual AirPods.

**Comment by d4rken** [comment]
See here for details from the MagicPods project:

* https://github.com/steam3d/MagicPodsCore/blob/master/src/aap/Aap.h
* https://github.com/steam3d/MagicPodsCore/blob/master/src/aap/Aap.cpp

and here is a python script (ty @fynngodau) that can connect to AirPods from a desktop computer:

```python
import bluetooth

# Change to MAC address of your AirPods
address = "MAC_ADDRESS"

aap_service = "74EC2172-0BAD-4D01-8F77-997B2BE0722A"
aap_port = 0x1001

# Commands taken from https://github.com/steam3d/MagicPodsCore/blob/master/src/aap/Aap.h
cmd_handshake = b"\x00\x00\x04\x00\x01\x00\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00"
cmd_off = b"\x04\x00\x04\x00\x09\x00\x0d\x01\x00\x00\x00"
cmd_on = b"\x04\x00\x04\x00\x09\x00\x0d\x02\x00\x00\x00"
cmd_transparency = b"\x04\x00\x04\x00\x09\x00\x0d\x03\x00\x00\x00"

services = bluetooth.find_service(address=address)

service = [s for s in services if s["service-classes"] == [aap_service]]

if not service:
    print("Device does not have AAP service")
    exit()

sock = bluetooth.BluetoothSocket(bluetooth.L2CAP)

sock.connect((address, aap_port))

print("Connected to AirPods")
print("Sending handshake...")

sock.send(cmd_handshake)

while True:
    # Ignore the responses because I am unsure what they mean
    # res = sock.recv(1024)
    # print(f"Response: {res.hex()}")

    print("Select command:")
    print("1. Turn off")
    print("2. Turn on")
    print("3. Toggle transparency")
    print("4. Exit")

    cmd = input("Enter command: ")

    if cmd == "1":
        sock.send(cmd_off)
    elif cmd == "2":
        sock.send(cmd_on)
    elif cmd == "3":
        sock.send(cmd_transparency)
    elif cmd == "4":
        break

sock.close()
```

**Comment by d4rken** [comment]
Enabling this could enable/fix #38,#186,#187,#191,#204,#212

**Comment by fynngodau** [comment]
> and here is a python script (ty @fynngodau)

I'm pretty sure I have nothing to do with that script, must have been somebody else :blush:

**Comment by vulpes2** [comment]
Tried playing with this and I didn't get very far unfortunately. Was able to establish the connection on a desktop computer too, but didn't manage to get anything to happen on Android 14.

The names for `createInsecureL2capChannel` and `createL2capChannel` are pretty misleading, if you look at the source closely, you can see they're for BLE only, not BR/EDR. The `0x1001` PSM was reported as invalid because BLE uses SPSM instead of PSM, which are only two bytes in size. It turns out both `createL2capSocket` and `createInsecureL2capSocket` are hidden APIs, and I didn't have xposed on my phone, so I couldn't get any further. Maybe some HCI packet logs will help figuring out why the connection is failing.

**Comment by d4rken** [comment]
Thanks for trying @vulpes2 :beers:. I had some hope that Android 15 would add more APIs here, but didn't see anything in the changelogs so far :frowning:

**Comment by vulpes2** [comment]
Seems like L2CAP on BR/EDR was never supported for 3rd party use on Android, which is kind of surprising. Not sure why did they make a new API to create L2CAP sockets for LE, but still wouldn't allow it over BR/EDR to this day.

**Comment by xerootg** [comment]
Any chance someone with an understanding of L2CAP (I'm certainly not an L2CAP understander) could do a PR into AOSP?

**Comment by vulpes2** [comment]
The function exists, you are just not allowed to use it from a normal app. Neither Android nor iOS allow you to create raw L2CAP connections on BR/EDR, but obviously both use them heavily at the system level.

**Comment by DragonSWDev** [comment]
@vulpes2 So it might be not possible at all unless you root your device?

**Comment by vulpes2** [comment]
Effectively yes, LSPosed requires Magisk.

**Comment by kavishdevar** [comment]
I have also been working on this, but have had no luck on android (unlike linux)... i am planning to publish my findings (and a small app for linux) that i found while reverse engineering using the packetlogger on mac (set ANC mode, etc.), but i think it's useless for now considering the psm value thing..

**Comment by vulpes2** [comment]
I've been poking at this for a little while now, and have managed to map out most of the packet formats, opcodes and configuration keys as of version `6F8`. Validating everything is not possible for me as I don't have access to the AirPods Max and Beats devices at the moment. Haven't had time to look into the recent `7A294` fw either, it has the new head gesture stuff which are not super critical for desktop use. When I have more time I'll write up some docs on the protocol, and provide a cross platform reference library to handle the connection logic.

**Comment by steam3d** [comment]
@vulpes2 I have fully implemented the work of all functions of all versions of airpods in MagicPods, it works for any firmware version. The main problem is that the functions are hardcoded, and not obtained dynamically as apple does.

Apple uses several configuration requests, but I still don't understand the logic behind it asking for them. It seems like it depends on the protocol version.

**Comment by vulpes2** [comment]
The capability stuff and notification type bitfield is a total nightmare to deal with, I have managed to decipher a few of them, but not enough to make a difference just yet. By configuration requests, do you mean opcode 0x9, or the initial capability response with opcode 0x2?

**Comment by kavishdevar** [comment]
@d4rken could you provide a simple apk with xposed with 3 simple toggles for ANC modes? or maybe even publish the code? I'll try it on my phone and try experimenting..

**Comment by kavishdevar** [comment]
i do have xposed installed, i can try someone else's app.. mine doesn't work, the `socket.connect()` takes for ever and doesn't return with an error or something.

Also, interestingly, this says public
![image](https://github.com/user-attachments/assets/ba58b054-f93d-4476-8b3e-68cc0aa9a543)

**Comment by kavishdevar** [comment]
https://github.com/adolfintel/OpenPods/issues/58#issuecomment-2368255376

TL;DR Got it working by commenting a few lines in the bluetooth stack's code for managing l2c connections (the ones which were blocking when sending data to the airpods, and getting no response). I made a small app, using the hidden method to create br/edr l2c socket. i can change listening mode, get battery data, and in-ear detection works, for now.

**Comment by steam3d** [comment]
I've started working on AAP protocol documentation in python, decided to abandon c++ in favor of simplicity.
https://github.com/steam3d/MagicPodsCore

**Comment by kavishdevar** [comment]
I've created a issue on issuetracker for this! https://issuetracker.google.com/issues/371713238. please upvote :)

**Comment by aikooo7** [comment]
Just want to say that this feature is a banger and could be a huge deal breaker for a lot of people to start using airpods in android, maybe even buying just to use on android

**Comment by kavishdevar** [comment]
@aikooo7 you can try a beta app if you have a rooted phone right now!

**Comment by aikooo7** [comment]
> @aikooo7 you can try a beta app if you have a rooted phone right now!

I do have root, where can I get the app?

**Comment by kavishdevar** [comment]
Are you comfortable with working with a shell environment?

If you are then, you will have to create a shell script in post-data-fs for overlaying the library file (I'll upload it on my [repo](https://github.com/kavishdevar/aln) later today) in /apex (where the bluetooth services live), and then another root module, or a script (ksu or magisk) to overlay the same library in /system/lib64 (because apex files have signature verification). Then you should be able to use my app (same repo)..

(The app works best for pro 2, and is very much a work in progress!!)

**Comment by kavishdevar** [comment]
@aikooo7,  instructions and files on my [first release](https://github.com/kavishdevar/aln/releases/tag/v0.0.1)

**Comment by K3NOXOFFICIAL** [comment]
> I've created a issue on issuetracker for this! https://issuetracker.google.com/issues/371713238. please upvote :)

@kavishdevar 
The issue on Issuetracker was just updated

**Comment by StupidRepo** [comment]
Considering we have that Magisk 'fix' for now, could we get some kind of implementation of what ALN has, into CAPod for the rooties? 😉

**Comment by d4rken** [comment]
> Considering we have that Magisk 'fix' for now, could we get some kind of implementation of what ALN has, into CAPod for the rooties? 😉

Root support is currently out of scope.
Patching the Bluetooth stack is very device specific and too brittle. Adding this would result in 5 people have it working and 95 complaining why it broke their Bluetooth.
Maybe a more general root solution becomes available or Google patches [371713238](https://issuetracker.google.com/issues/371713238) in a future update.

**Comment by StupidRepo** [comment]
> Root support is currently out of scope. Patching the Bluetooth stack is very device specific and too brittle. Adding this would result in 5 people have it working and 95 complaining why it broke their Bluetooth. Maybe a more general root solution becomes available or Google patches [371713238](https://issuetracker.google.com/issues/371713238) in a future update.

I guess I'm one of those 5 people lol. Anyway, thanks for the response 😁

**Comment by kavishdevar** [comment]
> google ain't doing shit about this... and i can't unlock my bootloader without losing all my data

very low chances they'd do anything. I’ll still consider it a bug because even if a company doesn't follow standards, it is still a responsiblity of the software to handle everything safely. There is a new stack (Gabeldorsche) that is in development since A13, but no idea when that is going to be in production, and if it suffers from the same issue as Fluoride. It’s just better to root your device because then you could also use 2 devices at the same time like many other headphones, use them as hearing aids regardless of your region support (possibly soon).

**Comment by Leclowndu93150** [comment]
oh well, i managed to get LSPosed running but i can't patch system apps apparently, i'm gonna cry.

i'll wipe my phone when i have the time

**Comment by grantsavill** [comment]
Amazing work. Appears to be working for me on a Pixel 10 Pro with the latest updates. Was the fix included in the April update?

**Comment by DragonSWDev** [comment]
Is it supposed to work on latest non beta Android? I have Pixel 9 with latest updates, I installed version 5.0.1 from GitHub but I can't find option to change ANC mode. Does it require premium?

**Comment by d4rken** [comment]
@grantsavill 

I think so, the latest A16 April update.

> Is it supposed to work on latest non beta Android? I have Pixel 9 with latest updates, I installed version 5.0.1 from GitHub but I can't find option to change ANC mode. Does it require premium?

@DragonSWDev Works on my Pixel 9 Pro with latest Android 16 update.
Some options require premium, but it should still be visible whether the connection worked.

Just BLE, no L2CAP socket connection (AAP) possible:

Card only shows a little broadcast antenna icon near "AirPods Pro" and the settings screen shows no options and only "connect".

<img width="250"  alt="Image" src="https://github.com/user-attachments/assets/41e9fabc-fa85-4bae-a50d-b2583d6dd0e0" />
<img width="250"  alt="Image" src="https://github.com/user-attachments/assets/cce92f29-c1aa-4c62-951b-135e3b480033" />


BLE + AAP:

Card shows Bluetooth icon near "AirPods Pro 2 USB-C" (+ BLE broadcast antenna, and IRK + ENC key icons as the app now automatically retrieves those via AAP. Settings are fully visible and multiple options shown.

<img width="250"  alt="Image" src="https://github.com/user-attachments/assets/344d6f1a-d5c6-4032-8e26-8269e370c057" />
<img width="250"  alt="Image" src="https://github.com/user-attachments/assets/5de74f14-d5c2-4fbb-be99-57e63d7a9ae5" />

**Comment by DragonSWDev** [comment]
@d4rken It's working now, I had to remove and add profile again. Thank you for your help and for your work.

**Comment by d4rken** [comment]
> [@d4rken](https://github.com/d4rken) It's working now, I had to remove and add profile again. Thank you for your help and for your work.

Got any more detail? This should not be required 🤔

**Comment by DragonSWDev** [comment]
@d4rken I had CAPod installed from Google Play Store, I've removed it and installed version 5.0.1-rc0 from GitHub. I wasn't able to change ANC mode (it looked exactly the same as in your first screenshot) but after I went to Settings, then to Devices, selected my device, removed it and added again (selecting my AirPods 2 from the list of paired devices) it started working. I noticed that you already merged fix for that, that was probably it as I had been using CAPod long before this feature was added.
